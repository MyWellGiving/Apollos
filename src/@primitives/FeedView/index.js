import React from 'react';
import {
  Text,
  View,
} from 'react-native';
import PropTypes from 'prop-types';
import { map } from 'lodash';
import {
  pure,
  compose,
  branch,
  renderComponent,
  componentFromProp,
} from 'recompose';

import MediaCard from '@primitives/MediaCard';
import FeedList from './FeedList';

const defaultFeedItemRenderer = ({ item }) => { // eslint-disable-line
  console.log(item.content.isLight);
  return (
    <MediaCard
      title={item.title}
      category={item.channelName}
      image={item.content.images.length ?
        `https:${item.content.images[0].url}` :
        'https://placeholdit.co//i/600x400?text=:`-( No Image In Array!'
      }
      cardColor={item.content.colors.length ? `#${item.content.colors[0].value}` : ''}
      isLight={item.content.isLight}
    />
  );
};

// TODO: replace with component from #115
const defaultEmptyStateRenderer = () => map([1, 2, 3], key => (
  <View key={key} style={{ height: 250, margin: 10, backgroundColor: 'rgba(0,0,0,0.1)' }}>
    <Text>...</Text>
  </View>
));

const renderEmptyState = renderComponent(componentFromProp('renderEmptyState'));

const enhance = compose(
  pure,
  branch(({ isLoading, feed }) => isLoading && !feed.length, renderEmptyState),
);

const FeedView = enhance(({
  isLoading,
  refetch,
  feed,
  fetchMore,
  ...otherProps
}) => (
  <FeedList
    {...otherProps}
    refreshing={isLoading}
    onRefresh={refetch}
    onEndReached={fetchMore}
    data={feed}
  />
));

FeedView.defaultProps = {
  isLoading: false,
  onEndReachedThreshold: 0.7,
  keyExtractor: item => item.entryId,
  feed: [],
  refetch: undefined,
  fetchMore: undefined,
  renderItem: defaultFeedItemRenderer,
  renderEmptyState: defaultEmptyStateRenderer,
};

FeedView.propTypes = {
  isLoading: PropTypes.bool,
  feed: PropTypes.array, // eslint-disable-line
  refetch: PropTypes.func,
  fetchMore: PropTypes.func,
  renderItem: PropTypes.func,
  renderEmptyState: PropTypes.func,
};

export default FeedView;
