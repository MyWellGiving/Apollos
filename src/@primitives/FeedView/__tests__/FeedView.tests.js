import React from 'react';
import renderer from 'react-test-renderer';

import ThemeProvider from '@primitives/ThemeProvider';
import FeedView from '../';

describe('The FeedView component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <FeedView
          feed={[
            {
              entryId: '1',
              title: 'hi',
              channelName: 'bob',
              content: {
                images: [{
                  url: 'https://placeholdit.co//i/600x400?text=:`-( No Image In Array!',
                }],
                colors: [],
              },
            },
            {
              entryId: '2',
              title: 'hi 2',
              channelName: 'bob 2',
              content: {
                images: [{
                  url: 'https://placeholdit.co//i/600x400?text=:`-( No Image In Array!',
                }],
                colors: [{
                  value: 'fff',
                }],
              },
            },
          ]}
        />
      </ThemeProvider>,
    );
    expect(tree).toMatchSnapshot();
  });
  it('renders empty state', () => {
    const tree = renderer.create(
      <FeedView
        isLoading
        feed={[]}
      />,
    );
    expect(tree).toMatchSnapshot();
  });
});
