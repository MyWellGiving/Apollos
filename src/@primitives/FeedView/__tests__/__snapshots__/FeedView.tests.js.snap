// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The FeedView component renders correctly 1`] = `
<RCTScrollView
  data={Array []}
  disableVirtualization={false}
  feed={
    Array [
      Object {
        "channelName": "bob",
        "entryId": "1",
        "title": "hi",
      },
      Object {
        "channelName": "bob 2",
        "entryId": "2",
        "title": "hi 2",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={undefined}
  onEndReachedThreshold={0.7}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onRefresh={undefined}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onViewableItemsChanged={undefined}
  refreshing={false}
  renderEmptyState={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  windowSize={21}
>
  <View />
</RCTScrollView>
`;

exports[`The FeedView component renders empty state 1`] = `
Array [
  <View
    style={
        Object {
            "backgroundColor": "rgba(0,0,0,0.1)",
            "height": 250,
            "margin": 10,
          }
    }
>
    <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
    >
        ...
    </Text>
</View>,
  <View
    style={
        Object {
            "backgroundColor": "rgba(0,0,0,0.1)",
            "height": 250,
            "margin": 10,
          }
    }
>
    <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
    >
        ...
    </Text>
</View>,
  <View
    style={
        Object {
            "backgroundColor": "rgba(0,0,0,0.1)",
            "height": 250,
            "margin": 10,
          }
    }
>
    <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
    >
        ...
    </Text>
</View>,
]
`;
